lib/tree.rb

```diff
@@ -1,64 +1,127 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current, key, value)
+    return TreeNode.new(key, value) if current.nil?
+    
+    if key < current.key
+      current.left = add_helper(current.left, key, value)
+    else
+      current.right = add_helper(current.right, key, value)
+    end 
+    return current
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n)
   def find(key)
-    raise NotImplementedError
+    return nil if @root == nil
+    return find_helper(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def find_helper(current, key)
+    return nil if current.nil?
+    return current.value if current.key == key
+    
+    if key < current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    list = []
+    return inorder_helper(@root, list)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current, list)
+    return list if current.nil?
+    
+    inorder_helper(current.left, list)
+    list << {key: current.key, value: current.value}
+    inorder_helper(current.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    list = []
+    return preorder_helper(@root, list)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current, list)
+    return list if current.nil?
+    
+    list << {key: current.key, value: current.value}
+    preorder_helper(current.left, list)
+    preorder_helper(current.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    list = []
+    return postorder_helper(@root, list)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(current, list)
+    return list if current.nil?
+    
+    postorder_helper(current.left, list)
+    postorder_helper(current.right, list)
+    list << {key: current.key, value: current.value}
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
This isn't working as it's always returning the height of the right side.

Consider this test

```ruby
      tree.add(60, "sam1")
      tree.add(50, "sam2")
      tree.add(40, "sam3")
      tree.add(30, "sam4")
      tree.add(20, "sam5")
      expect(tree.height()).must_equal 5
```